Nasal/engine.nas

```diff
@@ -143,6 +178,65 @@ var oil_consumption = maketimer(1.0, func {
     }
 });
 
+# Oil Refilling
+var oil_refill = func(){
+    var service_hours = getprop("/engines/active-engine/oil-service-hours");
+    var oil_level     = getprop("/engines/active-engine/oil-level");
+    var refilled      = oil_level - previous_oil_level;
+    #print("OIL Refill init: svcHrs=", service_hours, "; oil_level=",oil_level, "; previous_oil_level=",previous_oil_level, "; refilled=",refilled);
+    
+    if (refilled >= 0) {
+        # when refill occured, the new oil "makes the old oil younger"
+        var pct = 0;
+        if (oil_level > 0) {
+            pct = previous_oil_level / oil_level;
+        }
+        var newService_hours = service_hours * pct;
+        setprop("/engines/active-engine/oil-service-hours", newService_hours);
+        #print("OIL Refill: pct=", pct, "; service_hours=",service_hours, "; newService_hours=", newService_hours, "; previous_oil_level=", previous_oil_level, "; oil_level=",oil_level);
+    }
+    
+    previous_oil_level = oil_level;
+}
+
+# ======= Oil temperature jsbsim compensator =======
+# Currently, jsbsim oil temperature always initializes at 60°F.
+# We want an temperature that initialize to environment temperature until first start
+# and then gradually switch over to the real jsbsim value after some time.
+var calculate_real_oiltemp = maketimer(0.5, func {
+    if (!getprop("/engines/active-engine/already-started-in-session")) {
+        # engine is still cold
+        var temp_env        = getprop("/environment/temperature-degf") or 60;
+        var temp_jsbsim_oil = getprop("/engines/active-engine/oil-temperature-degf") or 60;
+        current_temp_diff   = temp_jsbsim_oil - temp_env;
+        setprop("/engines/active-engine/oil-temperature-env-diff", current_temp_diff);
+    } else {
+        # engine has been started at least one time:
+        # gradually remove the difference as jsbsim adapts to real environment temperature
+        calculate_real_oiltemp.stop();
+        interpolate("/engines/active-engine/oil-temperature-env-diff", 0, 180); # hand over to jsbsim caluclation gradually over 2 minutes
+    }
+});
+
+# ========== engine damage due to over-RPM =========
+# Initialize variables to reduce GC load
+var rpm = getprop("/engines/active-engine/rpm");
+var damage = 0.0;
+
+var eng_damage_function = maketimer(1.0, func {
+    if (getprop("/engines/active-engine/damage_allowed")) {
+        rpm = getprop("/engines/active-engine/rpm"); # we need to update the RPM, but we do not set the variable here to avoid GC load
+    
+        if (rpm > 2700) {
+            damage = damage + ((rpm - 2700) / 100); # 1 point of damage per 100RPM over limit added per second, so the damage increases faster the more you go over the limit
```

Comment:
So you get a linear increase of damage for the engine RPM above the threshold (2700 RPM)? I'm not completely sure about that. Why not have a slow increase for engine RPM near the threshold, but a much higher damage increase if you go far above the threshold? Exponential growth.

I think you could use a JSBSim/Autopilot table that defines the damage/second as a function of RPM-above-the-threshold. That is also easier to tune than having just a single formula. This way the data that controls how fast the damage increases (the table) is separated from the code that applies the damage (this timer function).

If you have a table, you could even use an integrator if that's what you prefer.

Reply:
I managed to get the 180 HP engine to 4600 RPM while diving, but that's ridiculously high. Perhaps create a table from 0 (2700 RPM) to 800 (3500 RPM).

Reply:
Thank you, I will attempt to implement your suggestions by Monday ;)

Reply:
I now have a rpm over redline function, and am now creating the table.